### ISSUE TITLE
Feature reques, please consider to replace zlib to 50% faster libdeflate

### ISSUE THREAD

**Issue by dmitriyse** [glennrp/libpng#205]
Preformance comparision can be obtained from this report http://www.htslib.org/benchmarks/zlib.html

**Comment by benkasminbullock** [comment]
It sounds interesting to add this to libpng but the API is not compatible so someone needs to fork the repository, add optional libdeflate support, then add a pull request back to libpng.

**Comment by zachlewis** [comment]
Just wanted to bump this and register additional interest. 

I've been exploring PNG performance in various projects w.r.t. the increasingly popular use of alternate zlib drop-in replacements, and the difference really is significant... moreover, the performance improvements really only significantly affect PNG, compared to other formats.

I'm [reluctant](https://github.com/AcademySoftwareFoundation/OpenImageIO/pull/4639) to ship a library that links PNG built against, say, compatibility-mode zlib-ng, because I'm not confident I can reliably anticipate or control for surprising behavior involving other parts of the library linking dependencies that themselves link vanilla zlib, or other projects linking our library that themselves link (or have other dependencies that link) vanilla zlib. 

I feel like PNG's write performance, vis a vis deflate use, is best addressed either directly in PNG, or in zlib itself.

If I could take this on myself, I would; but I'm not a C++ programmer. (Though I will at least attempt to dive in and see if I can figure out what needs figuring out).

Is there any chance I could convince pnggroup core contributors to look into, and perhaps take on, implementing libdeflate to improve write performance? 

I feel like this would do quite a bit to make things easier and safer for downstream projects concerned with PNG performance, without having to take steps that potentially affect behavior of other libraries.

**Comment by jbowler** [comment]
What write API did you use?

**Comment by zachlewis** [comment]
Hey John, I appreciate your attention, I know you're a busy guy.

tl;dr -- we're not using the simplified API.

Disclaimer: I wish I knew how to concisely and definitively answer your question, but I'm not a C++ programmer, I'm not the author of the PNG-facing portion of our code, and prior to your reply, I wasn't familiar enough with libpng to know that there's multiple write APIs in the first place! Please bear with me.

We support writing a variety of ancillary chunks (and compression schemes), so I don't think the simplified API is suitable for us. If functions beginning with `png_image` is indicative of the simplified API, we're not using any of those. (Looking at the manual, I see there's a `PNG_IMAGE_FLAG_FAST` flag, which is an extremely interesting option, one for which we could definitely find a use case; but we're primarily writing PNGs for a variety of distribution cases.)

If you (or anyone else) have the time and energy to provide guidance re: the APIs we're using, or what we could be doing better, we'd really appreciate it.

The library is called OpenImageIO, and it's been around for ~15 years give or take, so it's very possible there are newer and better ways of using libpng that we're not aware of! All our PNG-specific code is located here: https://github.com/AcademySoftwareFoundation/OpenImageIO/tree/main/src/png.imageio


(Also, quick question -- would it be possible for us to implement libdeflate as a potentially default compression scheme ourselves, without having to make changes to  libpng itself?)

**Comment by jbowler** [comment]
Well, from memory, OpenImageIO would have to be rewritten to use libdeflate.   I think I looked at the read code last time we talked but it was the lowest level API IRC and so I assume the write code probably is.  It certainly isn't the simplified API which has explicit support for fast write; the rest of the write code defaults to favouring **size** over **speed** and has to be explicitly tuned at the zlib level for large images.

The low level write API is not compatible with libdeflate; the write **API** is not compatible.  OpenImageIO would have to be written to use a different API.

**Comment by jbowler** [comment]
>Also, quick question -- would it be possible for us to implement libdeflate as a potentially default compression scheme ourselves, without having to make changes to libpng itself?

Given that it is OpenImageIO and it doesn't do much with libpng (IRC) the answer is **yes**.

I did a couple of quick tests with `iconvert` to verify what I remember and it does seem that OpenImageIO on **write** uses hardly any of the available PNG formats.

Pretty much it only writes 8 and 16-bit G, GA, RGB, RGBA; it even converts colour mapped images to RGB (and doesn't convert back!)

So it's ultra-non-compressing.

To use `libdeflate` you need to assemble the (uncompressed) data into a single big buffer.  Because OpenImageIO data is always byte aligned that is easy; just copy it into a big buffer with each row preceded by a single byte.  Set **that** byte to 0.  The data formats are strictly:

```
G
GA
RGB
RGBA
```
Plus the same formats with big-endian 16-bit values (i.e. most significant byte first).

You can then use `libdeflate`, but you can also use the corresponding `zlib` and `zlib-ng` APIs with `deflateBound` (as in the corresponding libdeflate function).

Note that `zlib` is somewhat more robust than `libdeflate`; if the output buffer is not big enough for the input you get a second shot, whereas `libdeflate` just fails.

Note also that the output buffer size is always bigger (typically 10% bigger) than the input.  Consequently this approach will really suck with big images; you end up allocating two enormous extra buffers, each bigger than the whole of the original image, and the code actually really does write to all of them.  This is what typically kills performance; it's absolutely the worst algorithm for large datasets.

_However `libdeflate` forces you to do this; with `zlib` you can choose the maximum size of the buffers._

You don't need `libpng` _at all_ for this, although you do need code to do the Ethernet CRC32 (it's only 8 or so lines of code, but it's a non-obvious 8 lines :-)

If you are prepared to use `libpng` you can do so to get the support for the "chunk" format; `png_write_chunk`.  This is the unterAPI; generally disregarded and treated as unimportant yet the whole of the foundation of the `libpng` write code.  Given that, apart from the image data itself, OpenImageIO prepares the data for each chunk (the chunks I've seen it write) `png_write_chunk` makes the whole thing very easy.

If you want to do this with the `libpng` uberAPIs you really can't; `libpng` always works row-by-row and this defeats what I term "flash compression"; doing it all in one shot.

The "flash" approach inherently avoids one complete copy of the data but, in fact, the cost of that is probably over-rated.  I've had a brief conversation with Mark about avoiding it (the copy) and I know how to do it without breaking the libpng streaming APIs.  It may be that there is an implementation out there already which does it (`zlib-ng` maybe?)

The trick is to retain the previously handled uncompressed data until it disappears from the LZ77 "window".   `libpng` adopts a different approach **by default** - it reduces the window so that it is as small as possible while remaining as large as the data - but this only avoids the allocation (IRC Mark's implementation does not touch the window bytes until the specific bytes are required.)

All the same if you analyse OpenImageIO using the standard timing/performance testing approach I suspect you will find the time goes in places outside `zlib` for small images.

**Comment by zachlewis** [comment]
Thanks again, John! This is quite a bit to chew on, and I really appreciate the detailed response. I think I have a much better understanding of why `zlib` is more appropriate than `libdeflate` for PNG's purposes. Given how tunable `zlib` is, it's not unreasonable to expect folks to season to taste for specific use-cases. 

> Pretty much it only writes 8 and 16-bit G, GA, RGB, RGBA; it even converts colour mapped images to RGB (and doesn't convert back!)
> 
> So it's ultra-non-compressing.

Are you referring to a means for converting RGB to a perceptually uniform color space for compression + encoding, and a data chunk (ICC? cICP?) to signal that the decoder should convert back to RGB?

> Note also that the output buffer size is always bigger (typically 10% bigger) than the input. Consequently this approach will really suck with big images; you end up allocating two enormous extra buffers, each bigger than the whole of the original image, and the code actually really does write to all of them. This is what typically kills performance; it's absolutely the worst algorithm for large datasets.

Interesting. I'm specifically concerned with performance when writing very large PNGs, so I imagine any speed gains that could be achieved with `libdeflate` would necessarily come at the cost of memory use. This makes me wonder why OpenEXR's use of `libdeflate` is as performant as it is. Maybe "scanlines" or "tiles" are compressed independently with smaller LZ77 "windows" concurrently or something? I don't even know if that makes sense.

Anyway, thanks again for your reply -- I've much to consider!

**Comment by jbowler** [comment]
> > Pretty much it only writes 8 and 16-bit G, GA, RGB, RGBA; it even converts colour mapped images to RGB (and doesn't convert back!)
> > So it's ultra-non-compressing.
> 
> Are you referring to a means for converting RGB to a perceptually uniform color space for compression + encoding, and a data chunk (ICC? cICP?) to signal that the decoder should convert back to RGB?

No, not in this instance though I think I might have mentioned that elsewhere.

The PNG data model is a frame buffer, pretty much a latish 1980's frame buffer.  Contrast with JPEG ('89) where the data model is a continuous tone image.  PNG can be used to compress text just by putting the literal byte values of the characters into a frame buffer format; it's completely ridiculous as an image but it is a valid PNG.  The PNG formats with pixel bit depths below 16, or 8 for grayscale, are not in any sense capable of encoding continuous tone images.  Arguably **lossless** encoding of continuous tone images requires 16-bits _per component_ RGB or gray because using 8-bits per component will loose data even for SDR images; 100:1 dynamic range with 1% accuracy.  8 bits, so 255 steps of 1% (1.01) gives a dynamic range of under 13:1 but 9 bits with 511 steps gives slightly over 161:1.

OpenImageIO seems to be designed for continuous tone images and/or the corresponding data represented as the human perception we call colour.  So it fairs badly when given something else such as a lower bit depth frame buffer (regardless of actual contents).

The important point, though, is that OpenImageIO really does seem to be handling continuous tone/perceptual data.  This means that lossless formats like, for example, the base JPEG-XL, are appropriate and that PNG handling is possible but probably quite inefficient compared to JPEG-XL.  I compared a JPEG-XL HDR image with the output of djxl to the corresponding PNG and the file size increased more than 10 times.

I added **emphasis** to this part of your message:

> I'm specifically concerned with performance when writing very large PNGs, so I imagine any speed gains that could be achieved with `libdeflate` would necessarily come at the cost of memory use. This makes me wonder why OpenEXR's use of `libdeflate` is as performant as it is. **Maybe "scanlines" or "tiles" are compressed independently with smaller LZ77 "windows" concurrently or something?** I don't even know if that makes sense.

The OpenEXR files I produce seem to be scanline based with single precision floating point; 32 bits per component, 16 bytes per pixel.  Massively over-precise.  Nevertheless tile layout is also available.  In any case I believe the format is always "random access", so it is possible to jump to a particular line or tile without decompressing the intermediate data:

https://openexr.com/en/latest/OpenEXRFileLayout.html#file-layout

This architecture requires the compression of individual blocks to be "free standing"; lower levels (finer detail) of the hierarchy may depend on higher levels but the bit-level compression of each block does not depend on anything else.  So, yes, libdeflate or zlib inflate(Z_FINISH) just works.  Likewise on write zlib deflate(Z_FINISH) works because the cost of over-allocation of the buffer is manageable.  Indeed a read-write operation on a file may naturally over-allocate anyway if writing to an existing block (that's a 2D memory management challenge.)

What I hadn't realised is that you are interested in the write of large images.  I had assumed the small images used in many AI applications like face recognition.  Frequently this just uses fixed size grayscale bitmaps but fast PNG compression can be used to temporarily hold a lot of these from face detection of, e.g., a crowd.

Large images are primarily challenging in terms of memory management.  This can be done by using PNG with per-scanline compression then reassembling a valid zlib format stream from the individual compressed scanlines.  It's just a restart marker followed by the compressed block with header and trailer skipped; the Adler32 is a bit tricky but in principle it can be handled.

This, random access, is an active discussion in the PNG standards group (W3C) for PNGv4 (i.e. not PNGv3, which is finished but not finally approved).  It's also a topic of interest to me but having looked at JPEG-XL I'm not sure what the real application of PNG is; doing it in PNG would seem to be a poor reinvention of a wheel.

**Comment by zachlewis** [comment]
You are a fount of knowledge, my friend. 

I apologize -- I should have been more clear about my intentions. I've been investigating reports re: OpenImageIO performing relatively poorly on Windows compared to other platforms when writing very large 16-bit _tonally continuous_ PNGs -- in the specific test case, I believe folks were trying to write a 16K frame sequence, with each frame taking ~20s. We found that swapping in `zlib-ng` for `zlib` reduced write times to ~8s. I think we only briefly tested with smaller frames, but the difference was comparatively negligible. 

I ended up in this thread because it occurred to me that maybe if libpng used libdeflate _instead_ of zlib entirely, I could avoid having to worry about the implications of building some parts of OpenImageIO against dependencies that link static compatibility-mode `zlib-ng`, and other parts against dependencies that link static vanilla `zlib`... I just figured if it worked for OpenEXR and libtiff, maybe there's something there... and now I far better understand the situation. The answer is, anything's possible, technically, with a scanline-based approach; but `zlib`, being more robust and more widely available, makes some serious sense for typical PNG use cases, especially since `zlib` provides a means for streaming data in chunks, rather than having to allocate the entire output buffer ahead of time... 

> OpenImageIO seems to be designed for continuous tone images and/or the corresponding data represented as the human perception we call colour. So it fairs badly when given something else such as a lower bit depth frame buffer (regardless of actual contents).

Very astute. OpenImageIO has its roots in VFX production, so it isn't really concerned with RGB buffers lower than 8-bit. Incidentally, we tend to reserve OpenEXR's 32f buffers for encoding CG AOVs where the precision actually matters, like surface normal, point position, motion vector, UV, depth passes, etc. For everything else, we're ideally allocating radiometrically linear relative exposure values to half-float (16f) buffers, maximizing precision around 18% grey. Which is why OpenEXR isn't really suitable for nonlinear encodings, compared to integer-based containers.

To your greater point about whether PNG is the most appropriate format for interchanging this kind of data OpenImageIO was designed to work with -- honestly, it's not a preferred production format; but PNGs are ubiquitous, both in the wild, and in assets interchanged among vendors and clients, so... 🤷. OIIO's format plugins are informed by production needs. And the more PNG improves, the more use-cases we'll find ourselves needing to support. For example, when we write Rec.2100 PQ-encoded tonally continuous imagery mastered for HDR devices, we'd want our PNG writer to automatically write the indicative 9-16-9 `cICP` chunk. But that's a whole other can of worms. 

> This, random access, is an active discussion in the PNG standards group (W3C) for PNGv4 (i.e. not PNGv3, which is finished but not finally approved). It's also a topic of interest to me but having looked at JPEG-XL **I'm not sure what the real application of PNG is; doing it in PNG would seem to be a poor reinvention of a wheel**.

(emphasis mine). Without having any real stake in PNG myself, I would tend to agree. 

> Large images are primarily challenging in terms of memory management. This can be done by using PNG with per-scanline compression then reassembling a valid zlib format stream from the individual compressed scanlines. It's just a restart marker followed by the compressed block with header and trailer skipped; the Adler32 is a bit tricky but in principle it can be handled.

This ultimately tells me what I need to know. It only makes sense to use libdeflate for very large images if we're not trying to compress the entire buffer in one shot, because we'd have to allocate the entire output buffer ahead of time. But since we do already have functions wrapping calls to png_write_row, maybe the structure is already in place for us to neatly hook in per-scanline compression with libdeflate, and assemble correctly after the fact, like you're describing. 

(Or maybe we'll just stick with zlib-ng after all...!)

Anyway, thank you again -- it's good to know what I'm considering trying to do is at least viable in theory! Very much appreciate your insight.

**Comment by jbowler** [comment]
> Or maybe we'll just stick with zlib-ng after all...!

zlib-ng should be a drop in replacement for zlib under libpng.  People have tested it (not me) with libpng 1.6 and I believe they did so on Windows.  The general wisdom seems to be that zlib-ng is faster in a lot of cases.  I found the other recent issue about write speed:

#498 

Since you are writing 16-bit components with continuous tone data it is **very** important to turn off PNG row filtering; `png_set_filter(png_ptr, 0/*method*/, PNG_FILTER_NONE)`.  The problem is that the filters are applied to the 2-byte component values a byte at a time and this produces what is basically noise in the least significant byte with all the filters!  This should be happening in the default setup but it does not in 1.6 (IRC I changed it in 1.7 but I wasn't allowed to change it in 1.6. :-(

You can hack libpng 1.6 to do this; look for `png_write_IHDR` in pngwutil.c and then find the assignments to `png_ptr->do_filter`.  IRC that's the only place the defaulting is handled (in 1.6).  Just add `|| png_ptr->bit_depth == 16` at the end of the test on line 844 (currently) which precedes the setting of `PNG_FILTER_NONE`:

https://github.com/pnggroup/libpng/blob/44f97f08d729fcc77ea5d08e02cd538523dd7157/pngwutil.c#L844

Doing it as a hack means that you can static link libpng 1.6 (or even install a DLL in an appropriate Windows assembly) and change the behaviour of OpenImageIO without changing OpenImageIO itself.  The change I'm suggesting is, so far as I know, good for everyone except maybe low colour count things like posters but they probably don't end up with 16-bit components.

**Comment by hmaarrfk** [comment]
Here is a patch that should get you to "hard depend" on zlib-ng without compatibility mode.

I think compatibility modes are good for experiments, but cause alot of uncertainty in deployments.

Before applying the patch, i would recommend you simply get rid of many of the ZLIB_VERNUM (see [this](https://github.com/pnggroup/libpng/blob/7c67f70da1db5f1d7a75cf518d8e5bd704ba89ce/pngstruct.h#L36)) if statements, since they are for a version of ZLIB released in 2012.
https://zlib.net/fossils/

[0003-Use-zlib-ng-instead-of-zlib.patch.txt](https://github.com/user-attachments/files/23287740/0003-Use-zlib-ng-instead-of-zlib.patch.txt)

**Comment by FabioPedretti** [comment]
> "hard depend" on zlib-ng

> get rid of many of the ZLIB_VERNUM

@hmaarrfk Maybe you could open a PR for both?

**Comment by hmaarrfk** [comment]
@FabioPedretti are you able to review such a PR? i would prefer a request to come from some1 that can, and I'm simply not familiar with the faces here.

**Comment by FabioPedretti** [comment]
@hmaarrfk no, sorry, was just a suggestion.

**Comment by Demonese** [comment]
I'm maintaining a zlib-ng fork that might be helpful.

libdeflate is actively maintained and looks better, but I need minizip-ng, so I went with zlib-ng.

https://github.com/Legacy-LuaSTG-Engine/libpng-zlib-ng

I'm thinking about how to abstract the deflate functionality so that we can freely choose the deflate library instead of being locked into a specific vendor.
